jmh-benchmarks/src/main/java/org/apache/kafka/jmh/timeline/TimelineHashMapBenchmark.java

```diff
@@ -44,33 +49,126 @@
+    public static class TimelineMapInput {
+        public SnapshotRegistry snapshotRegistry;
+        public TimelineHashMap<Integer, String> map;
+        public final List<Integer> keys = createKeys(NUM_ENTRIES);
+
+        @Setup(Level.Invocation)
+        public void setup() {
+            snapshotRegistry = new SnapshotRegistry(new LogContext());
+            map = new TimelineHashMap<>(snapshotRegistry, keys.size());
+
+            for (Integer key : keys) {
+                map.put(key, String.valueOf(key));
+            }
+
+            Collections.shuffle(keys);
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class TimelineMapSnapshotInput {
+        public SnapshotRegistry snapshotRegistry;
+        public TimelineHashMap<Integer, String> map;
+        public final List<Integer> keys = createKeys(NUM_ENTRIES);
+
+        @Setup(Level.Invocation)
+        public void setup() {
+            snapshotRegistry = new SnapshotRegistry(new LogContext());
+            map = new TimelineHashMap<>(snapshotRegistry, keys.size());
+
+            for (Integer key : keys) {
+                map.put(key, String.valueOf(key));
+            }
+
+            int count = 0;
+            for (Integer key : keys) {
+                if (count % 1_000 == 0) {
+                    snapshotRegistry.deleteSnapshotsUpTo(count - 10_000);
+                    snapshotRegistry.createSnapshot(count);
+                }
+                map.put(key, String.valueOf(key));
+                count++;
+            }
+
+            Collections.shuffle(keys);
+        }
+    }
+
+
     @Benchmark
     public Map<Integer, String> testAddEntriesInHashMap() {
-        HashMap<Integer, String> map = new HashMap<>(NUM_ENTRIES);
+        HashMap<Integer, String> map = new HashMap<>();
         for (int i = 0; i < NUM_ENTRIES; i++) {
             int key = (int) (0xffffffff & ((i * 2862933555777941757L) + 3037000493L));
             map.put(key, String.valueOf(key));
         }
+
+        return map;
+    }
+
+    @Benchmark
+    public scala.collection.immutable.HashMap<Integer, String> testAddEntriesInImmutableMap() {
+        scala.collection.immutable.HashMap<Integer, String> map = new scala.collection.immutable.HashMap<>();
+        for (int i = 0; i < NUM_ENTRIES; i++) {
+            int key = (int) (0xffffffff & ((i * 2862933555777941757L) + 3037000493L));
+            map = map.updated(key, String.valueOf(key));
+        }
+
         return map;
     }
 
     @Benchmark
     public Map<Integer, String> testAddEntriesInTimelineMap() {
         SnapshotRegistry snapshotRegistry = new SnapshotRegistry(new LogContext());
-        TimelineHashMap<Integer, String> map =
-            new TimelineHashMap<>(snapshotRegistry, NUM_ENTRIES);
+        TimelineHashMap<Integer, String> map = new TimelineHashMap<>(snapshotRegistry, 16);
         for (int i = 0; i < NUM_ENTRIES; i++) {
             int key = (int) (0xffffffff & ((i * 2862933555777941757L) + 3037000493L));
```

Comment:
Why are we doing these things?

Reply:
Hmm, I'd just generate the randoms during set-up and add them to an array.

Reply:
Done.